nixos/modules/services/networking/knot.nix

```diff
@@ -0,0 +1,94 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.knot;
+
+  configFile = pkgs.writeText "knot.conf" cfg.extraConfig;
+  socketFile = "/run/knot/knot.sock";
+
+  knotConfCheck = file: pkgs.runCommand "knot-config-checked"
+    { buildInputs = [ cfg.package ]; } ''
+    ln -s ${configFile} $out
+    knotc --config=${configFile} conf-check
+  '';
+in {
+  options = {
+    services.knot = {
+      enable = mkEnableOption "Knot authoritative-only DNS server";
+
+      extraArgs = mkOption {
+        type = types.listOf types.str;
+        default = [];
+        description = ''
+          List of additional command line paramters for knotd
+        '';
+      };
+
+      extraConfig = mkOption {
+        type = types.lines;
+        default = false;
+        description = ''
+          Extra lines to be added verbatim to knot.conf
+        '';
+      };
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.knot-dns;
+        description = ''
+          Which Knot DNS package to use
+        '';
+      };
+    };
+  };
+
+  config = mkIf config.services.knot.enable {
+    systemd.services.knot = {
+      unitConfig.Documentation = "man:knotd(8) man:knot.conf(5) man:knotc(8)";
+      description = cfg.package.meta.description;
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network.target" ];
+      after = ["network.target" ];
+
+      serviceConfig = {
+        Type = "notify";
+        ExecStart = "${cfg.package}/bin/knotd --config=${knotConfCheck configFile} --socket=${socketFile} ${concatStringsSep " " cfg.extraArgs}";
+        ExecReload = "${cfg.package}/bin/knotc --config=${knotConfCheck configFile} --socket=${socketFile} reload";
```

Comment:
Why not simplify at least the knotc line by using the wrapped one?
_I'll look at the whole expression a bit more, later; I'm upstream for knot-resolver._

Reply:
I briefly discussed with @mweinelt on how to do that since he wasn't sure. My recommended way would be changing it as below.

```patch
diff --git a/nixos/modules/services/networking/knot.nix b/nixos/modules/services/networking/knot.nix
index df117071630..9ac7763152c 100644
--- a/nixos/modules/services/networking/knot.nix
+++ b/nixos/modules/services/networking/knot.nix
@@ -13,6 +13,22 @@ let
     ln -s ${configFile} $out
     knotc --config=${configFile} conf-check
   '';
+  cli-wrappers = pkgs.stdenv.mkDerivation {
+    name = "knot-cli-wrappers";
+    buildInputs = [ pkgs.makeWrapper ];
+    buildCommand = ''
+      mkdir -p $out/bin
+      makeWrapper ${cfg.package}/bin/knotc "$out/bin/knotc" \
+        --add-flags "--config=${configFile}" \
+        --add-flags "--socket=${socketFile}"
+      makeWrapper ${cfg.package}/bin/keymgr "$out/bin/keymgr" \
+        --add-flags "--config=${configFile}"
+      for executable in kdig khost kjournalprint knsec3hash knsupdate kzonecheck
+      do
+          ln -s "${cfg.package}/bin/$executable" "$out/bin/$executable"
+      done
+    '';
+  };
 in {
   options = {
     services.knot = {
@@ -55,7 +71,7 @@ in {
       serviceConfig = {
         Type = "notify";
         ExecStart = "${cfg.package}/bin/knotd --config=${knotConfCheck configFile} --socket=${socketFile} ${concatStringsSep " " cfg.extraArgs}";
-        ExecReload = "${cfg.package}/bin/knotc --config=${knotConfCheck configFile} --socket=${socketFile} reload";
+        ExecReload = "${cli-wrappers}/bin/knotc reload";
         CapabilityBoundingSet="CAP_NET_BIND_SERVICE CAP_SETPCAP";
         AmbientCapabilities="CAP_NET_BIND_SERVICE CAP_SETPCAP";
         NoNewPrivileges=true;
@@ -69,24 +85,7 @@ in {
       };
     };
 
-   environment.systemPackages = let
-      cli-wrappers = pkgs.stdenv.mkDerivation {
-        name = "knot-cli-wrappers";
-        buildInputs = [ pkgs.makeWrapper ];
-        buildCommand = ''
-          mkdir -p $out/bin
-          makeWrapper ${cfg.package}/bin/knotc "$out/bin/knotc" \
-            --add-flags "--config=${configFile}" \
-            --add-flags "--socket=${socketFile}"
-          makeWrapper ${cfg.package}/bin/keymgr "$out/bin/keymgr" \
-            --add-flags "--config=${configFile}"
-          for executable in kdig khost kjournalprint knsec3hash knsupdate kzonecheck
-          do
-              ln -s "${cfg.package}/bin/$executable" "$out/bin/$executable"
-          done
-        '';
-      };
-    in [ cli-wrappers ];
+   environment.systemPackages = [ cli-wrappers ];
 
   };
 ```

Reply:
Applied, thanks!